File: test/fixtures/sample.ts
// Fixture exercising the TypeScript symbol shapes the extractor recognizes.

export const MAX_WIDGETS = 32;
const defaultLabel = "widget";

export type WidgetId = string;

export type Box<T> = { value: T };

export interface Widget {
  readonly id: WidgetId;
  width: number;
  resize(width: number): Widget;
}

export enum Mode {
  Compact,
  Wide,
}

export abstract class BaseStore<T> {
  protected readonly items: T[] = [];

  abstract persist(item: T): void;

  get size(): number {
    return this.items.length;
  }
}

export class Store extends BaseStore<Widget> {
  static empty = new Store();

  constructor(private readonly label: string = defaultLabel) {
    super();
  }

  persist(item: Widget): void {
    this.items.push(item);
  }

  set label_(value: string) {
    void value;
  }

  private describe = (item: Widget): string => `${this.label}:${item.id}`;

  async fetch(id: WidgetId): Promise<Widget> {
    return { id, width: 1, resize: (width) => ({ id, width, resize: this.noop }) };
  }

  private noop = (): never => {
    throw new Error("unreachable");
  };
}

export function build(id: WidgetId): Widget {
  function normalize(value: string): string {
    return value.trim();
  }

  return { id: normalize(id), width: 1, resize: () => build(id) };
}

export function* widgets(): Generator<Widget> {
  yield build("a");
}

export const resize = (widget: Widget, width: number): Widget => ({
  ...widget,
  width,
});

export declare function ambientBuild(id: WidgetId): Widget;

export namespace Widgets {
  export const UNIT: WidgetId = "unit";

  export function unit(): Widget {
    return build(UNIT);
  }
}

File: test/fixtures/sample.py
"""Fixture exercising the Python symbol shapes the extractor recognizes."""

from abc import ABC, abstractmethod
from contextlib import contextmanager
from dataclasses import dataclass

MAX_WIDGETS = 32
default_label = "widget"

type WidgetId = str


@dataclass
class Widget:
    name: str
    width: int = 1

    @property
    def area(self) -> int:
        return self.width

    @area.setter
    def area(self, value: int) -> None:
        self.width = value

    @staticmethod
    def unit() -> "Widget":
        return Widget(name="unit")

    @classmethod
    def from_name(cls, name: str) -> "Widget":
        return cls(name=name)


class Store(ABC):
    LIMIT = 8

    def __init__(self) -> None:
        self.items: list[Widget] = []

    @abstractmethod
    def persist(self, widget: Widget) -> None:
        ...

    async def fetch(self, widget_id: WidgetId) -> Widget:
        return Widget(name=widget_id)


def build(name: str) -> Widget:
    def normalize(value: str) -> str:
        return value.strip()

    return Widget(name=normalize(name))


async def build_all(names: list[str]) -> list[Widget]:
    return [build(name) for name in names]


def identity[T](value: T) -> T:
    return value


@contextmanager
def opened(path: str):
    yield path


class TestWidget:
    def test_area(self) -> None:
        assert Widget(name="a").area == 1


@pytest.mark.parametrize("name", ["a", "b"])
def test_build(name: str) -> None:
    assert build(name).name == name
